src/cast_core/src/actors/CrudActor.cpp

```diff
@@ -0,0 +1,615 @@
+                    "'replaceOne' expects 'Filter' and 'Replacement' fields.");
+            }
+            auto filterTemplate = value_generators::makeDoc(filter, rng);
+            bsoncxx::builder::stream::document filterDocument{};
+            filterTemplate->view(filterDocument);
+            auto replacementTemplate = value_generators::makeDoc(replacement, rng);
+            bsoncxx::builder::stream::document replacementDocument{};
+            replacementTemplate->view(replacementDocument);
+            BOOST_LOG_TRIVIAL(info) << "filter: " << bsoncxx::to_json(filterDocument.view());
+            BOOST_LOG_TRIVIAL(info)
+                << "replacement: " << bsoncxx::to_json(replacementDocument.view());
+            _writeOps.push_back(mongocxx::model::replace_one(filterDocument.extract(),
+                                                             replacementDocument.extract()));
+        } else if (writeCommand == "updateOne") {
+            auto filter = writeOp["Filter"];
+            auto update = writeOp["Update"];
+            if (!filter || !update) {
+                throw InvalidConfigurationException(
+                    "'updateOne' expects 'Filter' and 'Update' fields.");
+            }
+            auto filterTemplate = value_generators::makeDoc(filter, rng);
+            bsoncxx::builder::stream::document filterDocument{};
+            filterTemplate->view(filterDocument);
+            auto updateTemplate = value_generators::makeDoc(update, rng);
+            bsoncxx::builder::stream::document updateDocument{};
+            updateTemplate->view(updateDocument);
+            _writeOps.push_back(
+                mongocxx::model::update_one(filterDocument.extract(), updateDocument.extract()));
+        } else if (writeCommand == "updateMany") {
+            auto filter = writeOp["Filter"];
+            auto update = writeOp["Update"];
+            if (!filter || !update) {
+                throw InvalidConfigurationException(
+                    "'updateMany' expects 'Filter' and 'Update' fields.");
+            }
+            auto filterTemplate = value_generators::makeDoc(filter, rng);
+            bsoncxx::builder::stream::document filterDocument{};
+            filterTemplate->view(filterDocument);
+            auto updateTemplate = value_generators::makeDoc(update, rng);
+            bsoncxx::builder::stream::document updateDocument{};
+            updateTemplate->view(updateDocument);
+            _writeOps.push_back(
+                mongocxx::model::update_many(filterDocument.extract(), updateDocument.extract()));
+        }
+    }
+
+    void run(const mongocxx::client_session& session) override {
+        auto bulk = (_onSession) ? _collection.create_bulk_write(session, _options)
+                                 : _collection.create_bulk_write(_options);
+        for (auto&& writeOp : _writeOps) {
+            bulk.append(writeOp);
+        }
+        auto result = bulk.execute();
+        if (!result) {
```

Comment:
It looks like [the `stdx::optional<result::bulk_write>` is only disengaged (i.e. `== std::nullopt`) when using unacknowledged writes](https://github.com/mongodb/mongo-cxx-driver/blob/r3.4.0/src/mongocxx/bulk_write.cpp#L175-L182). I think this if-statement is safe to remove.